kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,616 @@
+        '/', '.')
+
+    val hasCompanionObject = classProto.hasCompanionObjectName()
+    // todo allow custom constructor
+    val protoConstructor = classProto.constructorList
+        .single { it.isPrimary }
+    val constructorJvmSignature = protoConstructor.getJvmConstructorSignature(nameResolver,
+        classProto.typeTable)
+    val constructor = classProto.fqName
+        .let(nameResolver::getString)
+        .replace('/', '.')
+        .let(elementUtils::getTypeElement)
+        .enclosedElements
+        .mapNotNull { it.takeIf { it.kind == ElementKind.CONSTRUCTOR }?.let { it as ExecutableElement } }
+        .single { it.jvmMethodSignature == constructorJvmSignature }
+    val parameters = protoConstructor
+        .valueParameterList
+        .mapIndexed { index, valueParameter ->
+          val paramName = nameResolver.getString(valueParameter.name)
+
+          val nullable = valueParameter.type.nullable
+          val paramFqcn = valueParameter.type.extractFullName()
+              .replace("`", "")
+              .removeSuffix("?")
+
+          val actualElement = constructor.parameters[index]
+
+          val serializedName = actualElement.getAnnotation(Json::class.java)?.name
+              ?: paramName
+
+          val jsonQualifiers = actualElement.annotationMirrors
+              .filter { it.annotationType.getAnnotation(JsonQualifier::class.java) != null }
+
+          Property(
+              name = paramName,
+              fqClassName = paramFqcn,
+              serializedName = serializedName,
+              hasDefault = valueParameter.declaresDefaultValue,
+              nullable = nullable,
+              typeName = valueParameter.type.asTypeName(nameResolver, classProto::getTypeParameter),
+              jsonQualifiers = jsonQualifiers)
+        }
+
+
+    return Adapter(
+        fqClassName = fqClassName,
+        packageName = packageName,
+        propertyList = parameters,
+        originalElement = element,
+        hasCompanionObject = hasCompanionObject,
+        visibility = classProto.visibility!!,
+        elementUtils = elementUtils)
+  }
+
+  private fun errorMustBeDataClass(element: Element) {
+    messager.printMessage(ERROR,
+        "@${MoshiSerializable::class.java.simpleName} can't be applied to $element: must be a Kotlin data class",
+        element)
+  }
+
+  private fun Adapter.generateAndWrite(): Boolean {
+    val adapterName = "${name}_JsonAdapter"
+    val outputDir = generatedDir ?: mavenGeneratedDir(adapterName)
+    val fileBuilder = FileSpec.builder(packageName, adapterName)
+    generate(adapterName, fileBuilder)
+    fileBuilder
+        .build()
+        .writeTo(outputDir)
+    return true
+  }
+
+  private fun mavenGeneratedDir(adapterName: String): File {
+    // Hack since the maven plugin doesn't supply `kapt.kotlin.generated` option
+    // Bug filed at https://youtrack.jetbrains.com/issue/KT-22783
+    val file = filer.createSourceFile(adapterName).toUri().let(::File)
```

Comment:
I was thinking about this more yesterday and realized that this only needs to be called once.  KotlinPoet takes care of the fqcn path, so this is going to return the same file every time.

The only time that might be different is for different source sets (test, main, etc), but I think those will get different processors (please confirm), so that shouldn't matter.

Reply:
Not sure I follow. Could you elaborate on the "called once" bit? Since we're only calling it "once" per file anyway 🤔

Reply:
We call it once per file, but we're getting the root source directory from it, so it only needs to be called once per source set. KotlinPoet takes care of the package folders.